File: CokeLee777/src/com/leetcode/hashtable/SortCharactersByFrequency.java
package com.leetcode.hashtable;

import java.util.*;

/**
 * https://leetcode.com/problems/sort-characters-by-frequency/
 */
public class SortCharactersByFrequency {

    public static class MySourceCode{

        class Alphabet implements Comparable<Alphabet>{
            private final char letter;
            private final int count;

            public Alphabet(char letter, int count) {
                this.letter = letter;
                this.count = count;
            }

            public char getLetter() {
                return letter;
            }

            public int getCount() {
                return count;
            }

            @Override
            public int compareTo(Alphabet o) {
                if(this.count > o.count) return -1;
                else return 1;
            }
        }

        public String frequencySort(String s) {
            //알파벳의 개수 세기
            Map<Character, Integer> charToCount = new HashMap<>();
            for(char x : s.toCharArray()){
                charToCount.put(x, charToCount.getOrDefault(x, 0) + 1);
            }
            //리스트에 삽입 후 나온 개수가 큰 순서대로 정렬
            List<Alphabet> alphabets = new ArrayList<>();
            for (Character x : charToCount.keySet()) {
                alphabets.add(new Alphabet(x, charToCount.get(x)));
            }

            Collections.sort(alphabets);

            //정렬된 문자들 개수만큼 반복해서 저장
            StringBuilder ans = new StringBuilder();
            for (Alphabet alphabet : alphabets) {
                char letter = alphabet.getLetter();
                int count = alphabet.getCount();
                //문자의 개수만큼 반복
                ans.append(String.valueOf(letter).repeat(count));
            }

            return ans.toString();
        }
    }

    public static class Solution{

        public String frequencySort(String s) {
            Map<Character, Integer> map = new HashMap<>();
            for (char c : s.toCharArray())
                map.put(c, map.getOrDefault(c, 0) + 1);

            List<Character> [] bucket = new List[s.length() + 1];
            for (char key : map.keySet()) {
                int frequency = map.get(key);
                if (bucket[frequency] == null) bucket[frequency] = new ArrayList<>();
                bucket[frequency].add(key);
            }

            StringBuilder sb = new StringBuilder();
            for (int pos = bucket.length - 1; pos >= 0; pos--)
                if (bucket[pos] != null)
                    for (char c : bucket[pos])
                        for (int i = 0; i < pos; i++)
                            sb.append(c);

            return sb.toString();
        }
    }
}

File: dongdaang/hash_table/Sort_Characters_By_Frequency.py
#https://leetcode.com/problems/sort-characters-by-frequency/



#####################Counter 사용 풀이####################
from collections import Counter

class Solution:
    def frequencySort(self, s):
        ans = ''
        letters = Counter(s)        #카운터 함수로 알파벳 개수를 세주기
        letters = sorted(letters.items(), reverse= True, key= lambda x : x[1])      #letters 딕셔너리를 value 값을 기준으로 내림차순 정렬
        for i in letters:
            ans += i[0] * i[1]      #해당 글자를 카운트된 횟수만큼 출력하기
        return ans
        
a = Solution()
print(a.frequencySort("tree"))


####################Counter 사용 안한 풀이################
class Solution:
    def frequencySort(self, s):
        letters = {}
        ans = ''
        for c in s:                 #카운트 함수를 대신해 직접 카운트 해서 딕셔너리에 저장
            if c not in letters:
                letters[c] = 1
            else:
                letters[c] += 1
        letters = sorted(letters.items(), reverse= True, key= lambda x : x[1])
        for i in letters:
            ans += i[0] * i[1]
        return ans